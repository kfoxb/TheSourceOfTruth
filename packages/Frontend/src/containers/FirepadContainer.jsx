import React, { Component } from 'react';
import { database } from 'firebase';
import PropTypes from 'prop-types';
import CodeMirror from 'codemirror';
import { connect } from 'react-redux';
import Firepad from '../components/Firepad';
import { getCollection, getDocumentId } from '../helpers/firestore';

global.CodeMirror = CodeMirror;
const { fromCodeMirror } = require('firepad/dist/firepad');

class FirepadContainer extends Component {
  static propTypes = {
    history: PropTypes.shape({
      replace: PropTypes.func.isRequired,
    }).isRequired,
    isAuthenticated: PropTypes.bool.isRequired,
    match: PropTypes.shape({
      url: PropTypes.string.isRequired,
      params: PropTypes.shape({
        id: PropTypes.string.isRequired,
      }).isRequired,
    }).isRequired,
    readOnly: PropTypes.bool,
    uid: PropTypes.string,
  }

  static defaultProps = {
    readOnly: true,
    uid: '',
  }

  constructor(props) {
    super(props);
    this.state = {
      collection: getCollection(props.match.url),
      dialogIsOpen: false,
      documentId: getDocumentId(props.match.params.id),
      title: '',
    };
    this.setRef();
    this.ref.once('value', (snapshot) => {
      this.setState({ title: snapshot.val().title });
    });
  }

  componentDidMount() {
    if (this.props.isAuthenticated && this.props.uid) {
      this.initFirepad();
    }
  }

  componentDidUpdate(prevProps) {
    if (!prevProps.isAuthenticated && this.props.isAuthenticated && this.props.uid) {
      this.initFirepad();
    }
  }

  setRef() {
    const collectionRef = database().ref(this.state.collection);
    if (this.state.documentId) {
      this.ref = collectionRef.child(this.state.documentId);
    } else {
      const docRef = collectionRef.push();
      this.props.history.replace(`/${this.state.collection}/edit/${docRef.key}`);
      this.ref = docRef;
    }
  }

  initFirepad() {
    const { readOnly, uid } = this.props;
    const cm = CodeMirror(document.getElementById('firepad-container'), {
      lineWrapping: true,
      readOnly: readOnly ? 'nocursor' : false,
    });
    fromCodeMirror(
      this.ref, cm,
      {
        richTextToolbar: !readOnly,
        richTextShortcuts: !readOnly,
        userId: uid,
      },
    );
  }

  handleClose = () => {
    this.setState({ dialogIsOpen: false });
  };

  handleSubmit = () => {}

  openDialog = () => {
    this.setState({ dialogIsOpen: true });
  }

  handleTitleChange = ({ target: { value } }) => {
    const newTitle = { title: value };
    this.setState(newTitle, () => {
      this.ref.update(newTitle);
    });
  }

  render() {
    if (!this.props.isAuthenticated) {
      return (<p>Loading...</p>);
    }
    return (
      <Firepad
        title={this.state.title}
        handleClose={this.handleClose}
        handleSubmit={this.handleSubmit}
        handleTitleChange={this.handleTitleChange}
        dialogIsOpen={this.state.dialogIsOpen}
        openDialog={this.openDialog}
        readOnly={this.props.readOnly}
      />
    );
  }
}

const mapStateToProps = state => ({
  uid: state.user.uid,
  isAuthenticated: state.user.isAuthenticated,
});

export default connect(mapStateToProps)(FirepadContainer);
