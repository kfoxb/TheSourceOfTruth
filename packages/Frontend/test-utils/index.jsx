/* eslint-env jest */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow } from 'enzyme';

export const createAssertWithPropsToMatchSnapshot = (Component, defaultProps) =>
  (extraProps) => {
    const props = { ...defaultProps, ...extraProps };
    const wrapper = shallow(<Component {...props} />);
    expect(wrapper).toMatchSnapshot();
  };
