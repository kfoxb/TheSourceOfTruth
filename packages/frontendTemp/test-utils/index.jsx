/* eslint-env jest */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow } from 'enzyme';

export const makeShallowCreateWrapper = (Component, defaultProps) =>
  (extraProps) => {
    const props = { ...defaultProps, ...extraProps };
    return shallow(<Component {...props} />);
  };

export const createAssertWithPropsToMatchSnapshot = (Component, defaultProps) => {
  const createWrapper = makeShallowCreateWrapper(Component, defaultProps);
  return (extraProps) => {
    expect(createWrapper(extraProps)).toMatchSnapshot();
  };
};
