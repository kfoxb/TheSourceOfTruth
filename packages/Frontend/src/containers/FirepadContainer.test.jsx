/* eslint-env jest */
import { createAssertWithPropsToMatchSnapshot, makeShallowCreateWrapper } from '../../test-utils';
import FirepadContainer from './FirepadContainer';

describe('FirepadContainer', () => {
  const defaultProps = {
    history: {
      replace: () => {},
    },
    isAuthenticated: false,
    match: {
      url: '',
      params: {
        id: '',
      },
    },
  };

  const assertWithPropsToMatchSnapshot =
    createAssertWithPropsToMatchSnapshot(FirepadContainer.WrappedComponent, defaultProps);

  it('should render loading when not authenticated', () => {
    assertWithPropsToMatchSnapshot();
  });

  it('should render when authenticated', () => {
    assertWithPropsToMatchSnapshot({ isAuthenticated: true });
  });

  describe('journals', () => {
    const createWrapper = makeShallowCreateWrapper(FirepadContainer.WrappedComponent, defaultProps);

    it('should set it\'s collection to journals when the url contains journals', () => {
      const extraProps = {
        match: {
          url: '/journals/edit/',
          params: {
            id: '',
          },
        },
      };

      const wrapper = createWrapper(extraProps);
      expect(wrapper.state().collection).toEqual('journals');
    });
  });
});
