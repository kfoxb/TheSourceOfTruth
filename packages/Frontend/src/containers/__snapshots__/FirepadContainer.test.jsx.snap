// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirepadContainer should pass loading true props when not authenticated 1`] = `
<Firepad
  changingPhase={false}
  dialogIsOpen={false}
  error={false}
  handleClose={[Function]}
  handleSubmit={[Function]}
  handleTitleChange={[Function]}
  loading={true}
  notFound={false}
  openDialog={[Function]}
  readOnly={false}
  title=""
/>
`;

exports[`FirepadContainer should render when authenticated 1`] = `
<Firepad
  changingPhase={false}
  dialogIsOpen={false}
  error={false}
  handleClose={[Function]}
  handleSubmit={[Function]}
  handleTitleChange={[Function]}
  loading={true}
  notFound={false}
  openDialog={[Function]}
  readOnly={false}
  title=""
/>
`;
